DEF-3860 Faster Lua Check* functions

danengelbrecht-king opened the pull request:
Instead of doing a deep compare of the user data meta table we now add a special key to the meta table with a value of the type name hash and just validate that the key exists and has the correct value instead.

Removed pointless Lua gc methods and redundant type checks

PHYSICS_CONTEXT_NAME, INSTANCE_NAME, SCRIPT_CONTEXT has been complemented with corresponding hash keys in global lua table for faster lookup.

**CAUTION**: If you lock down the global namespace you need white-list the keys:
**`"__script_context"`** and **`2744634527`**
**`"__dm_script_instance__"`** and **`3700146495`**
**`"__PhysicsContext"`** and **`255264655`**

**CAUTION**: The equality comparison for Vector3, Vector4, Matrix, Quat and Hash has changed and will now return false when comparing two items with different types instead of generating a Lua type error. This is more in line with Luas built in types.

danengelbrecht-king committed:
Use meta table type hash instead of comparing full meta tables

Removed pointless Lua gc methods and redundant type checks

danengelbrecht-king committed:
Happy path for hash equality

danengelbrecht-king committed:
Dont do type check when we know the type already

danengelbrecht-king committed:
Use type hash for streams

danengelbrecht-king committed:
Cleanups and comments

danengelbrecht-king committed:
Fixed stack comments in GetUserType

danengelbrecht-king committed:
Changed global keys from string to hash

PHYSICS_CONTEXT_NAME, INSTANCE_NAME, SCRIPT_CONTEXT

danengelbrecht-king committed:
Timer uses hash for context key

danengelbrecht-king committed:
Use luaL_callmeta in dmScript::GetURL

danengelbrecht-king committed:
User types “__eq” returns false if types does not match

danengelbrecht-king committed:
No need to validate URL returned by META_TABLE_GET_URL callback

danengelbrecht-king committed:
Fixed bad character in gui_script.cpp

danengelbrecht-king committed:
Merge remote-tracking branch 'origin/dev' into DEF-3860-lua-check-with-hash

danengelbrecht-king committed:
Spelling fix

mathiaswking reviewed (approved):
Looks great.
Some minor fix needed.

danengelbrecht-king committed:
Review fixes

danengelbrecht-king committed:
Set both string name and hash for globals

To keep compatibility with users code that rely on it.
Also hid SCRIPT_CONTEXT_HASH inside script.cpp and added GetScriptContext()

danengelbrecht-king committed:
Use DM_LUA_ERROR if using DM_LUA_STACK_CHECK

danengelbrecht-king committed:
Merge remote-tracking branch 'origin/dev' into DEF-3860-lua-check-with-hash

danengelbrecht-king committed:
Fix SCRIPT_METATABLE_TYPE_HASH_KEY calculation

danengelbrecht-king committed:
Merge remote-tracking branch 'origin/dev' into DEF-3860-lua-check-with-hash
